Implement the `coredump` test check plugin

martinhoyer opened the pull request:
Leaving kdump and abrt to subsequent, separate PRs.

This PR is adding the initial implementation of a `coredump` Test Check.

Pull Request Checklist

* [x] implement the feature
* [x] write the documentation
* [x] extend the test coverage
* [x] update the specification
* [x] adjust plugin docstring
* [x] modify the json schema
* [x] mention the version
* [x] include a release note

martinhoyer commented:
Updated description, rebased to main. Marking as ready for review as I'd really appreciate some feedback.  

@happz continuing the discussion from chat - this is the only solution I found for `GuestCapabilities` and `_query_capabilities`, with which I can sleep at night...I hope it's ok for you too, once you check it *after* your PTO :)

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -285,14 +285,62 @@ class CheckRsyncOutcome(enum.Enum):
+    These values are stable across kernel versions as they are part of the kernel's
+    ABI guarantee. New capabilities can only be added with new numbers at the end.
+
+    See: include/uapi/linux/capability.h in the Linux kernel source.
```
We have an actual, official link: https://github.com/torvalds/linux/blob/master/include/uapi/linux/capability.h

martinhoyer replied:
Addressed in 62ad828b667fc6680072f375e6bb30edff9725a9

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -317,31 +365,43 @@ class GuestFacts(SerializableContainer):
+    capabilities: Optional[list[GuestCapability]] = field(
+        default=None,
+        serialize=lambda caps: [cap.value for cap in caps] if caps else None,
+        unserialize=lambda serialized: [GuestCapability(value) for value in serialized]
```
Just a nitpicking: IIUIC the code here, it'd emit a list of integers, which wouldn't be nice to read by humans. Maybe we should emit the names instead of values here, it would be a bit more work when unserializing, but inspecting the on-disk files would be easier for puny humans.

martinhoyer replied:
I've initially had a `Final[list[str]]`, but creating a type, `Literal[*<the final list>]` is not possible on Python <3.10 and while `Literal[tuple(<final list>)] does work, mypy can't deal with it.  Using `eumerate` didn't feel right to me, as it's a bit of black box, but I'd be happy to implement whatever you say is the best way @happz :)

happz replied:
I'm not sure I can catch up with the last comment. My point was to emit `cap.name` instead of `cap.value`: in files, we would see nice `CAP_FOO` instead of `79`. Unsrialize would need to do a bit more work, `GuestCapability.__members__[value]`.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -317,31 +365,43 @@ class GuestFacts(SerializableContainer):
+
+        return True
+
+    def has_capabilities(self, caps: list[GuestCapability]) -> bool:
```
All we need is an iterable, so `*caps: GuestCapability` would give us a tuple, and we could call this method without constructing a list, `has_capability(GuestCapability.CAP_SYSLOG, GuestCapability.CAP_PERFMON)`.

And then you can drop the `has_capability` method because it would be the same as `has_capabilities(a_single_capability)` :)

martinhoyer replied:
Addressed in 62ad828b667fc6680072f375e6bb30edff9725a9

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -579,13 +651,38 @@ def _query_is_ostree(self, guest: 'Guest') -> Optional[bool]:
+        capabilities: list[GuestCapability] = []
+        for cap in GuestCapability:
+            if bitmask & (1 << cap.value):
+                capabilities.append(cap)
```
List comprehension to the rescue! Or at least to a slightly shorter & more elegant oneliner without an accumulator variable:

```python
return [cap for cap in GuestCapability if bitmask & (1 << cap.value)]
```

martinhoyer replied:
Weird that wasn't caught by ruff.

martinhoyer replied:
Addressed in 62ad828b667fc6680072f375e6bb30edff9725a9

happz commented on tmt/steps/provision/podman.py:
```diff
@@ -475,11 +474,6 @@ def go(self, *, logger: Optional[tmt.log.Logger] = None) -> None:
-        # TODO: this might be allowed with `--privileged`...
-        self._guest.facts.capabilities[GuestCapability.SYSLOG_ACTION_READ_ALL] = False
-        # ... while this seems to be forbidden completely.
-        self._guest.facts.capabilities[GuestCapability.SYSLOG_ACTION_READ_CLEAR] = False
```
I'm not sure we should give up the distinction between these two "capabilities". Yes, it's a single capability, `CAP_SYSLOG`, but what I tried to emulate by hijacking the capabilities enum is the different level of functionality allowed: a privileged container seems to have a limited console ring buffer access, and it is possible to read the messages, but not remove them. That is something I would like to preserve, it would give users with container if not a clear picture of "this is what happened while the test was running", but at least a "this is all that has happened, sorry, can't do better".

I think it's no longer practical to squeeze this into `GuestCapabilities`, but having a way of tracking the "can read" vs "can read and erase" flavor of `CAP_SYSLOG` would be nice. I'll try to find a test that would demonstrate this behavior.

martinhoyer replied:
ok, looks like I misunderstood your chat message earlier. 
So if I understand it correctly, we could deduce this based on is-container, is-rootful, is-root / has-capability? (or something like that)

happz replied:
> So if I understand it correctly, we could deduce this based on is-container, is-rootful, is-root / has-capability? (or something like that)

Maybe. I kind of liked having it as a "capability", even an invented one, since `SYSLOG_ACTION_READ_ALL` and `SYSLOG_ACTION_READ_CLEAR` are genuine kernel buffer commands (https://linux.die.net/man/3/klogctl), so it's something that exists and is documented. Plus, the container plugin is aware of the situation already, it would be in better position to change these flags, the rest of tmt would just use them instead of inferring from other flags what to do.

So, as of now, the best solution to me seems to be adding them back to `GuestCapability` enum, with some high value. Basically reverting this bit here.

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,350 @@
+JournalSizeMax=767M
+MaxUse=0
+KeepFree=0
+"""
```
I would make this a module-level string, with some nice name. It's pretty much a multiline constant, method indentation doesn't suit it.

martinhoyer replied:
Addressed in c0c05df176604cf7fa3ec96d92f952e03b886726

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,350 @@
+                    "sh",
+                    "-c",
+                    f"mkdir -p /etc/systemd/coredump.conf.d && "
+                    f"echo '{config}' > /etc/systemd/coredump.conf.d/50-tmt.conf",
```
Check out `DEFAULT_SHELL`, I believe it's the one we should use for running shell commands.

BTW `ShellScript` might be more readable, `&&` plus a missing comma seems like a trap. How about this?

```python
guest.execute(ShellScript(
    f"""
    mkdir -p /etc/systemd/coredump.conf.d && "echo '{config}' > /etc/systemd/coredump.conf.d/50-tmt.conf
    """)

# This should also work:
guest.execute(
    ShellScript('mkdir -p /etc/systemd/coredump.conf.d')
    and ShellScript(f"echo '{config}' > /etc/systemd/coredump.conf.d/50-tmt.conf"))
```

`guest.execute()` should transparently convert shell script into command, with`DEFAULT_SHELL -c ...`

martinhoyer replied:
I've used ShellScript everywhere, except where it would require shell=True, at least as far as I understand.

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,350 @@
+                    "sh",
+                    "-c",
+                    f"mkdir -p {DEFAULT_TMP_PATH} && "
+                    f"date '+%Y-%m-%d %H:%M:%S' > {DEFAULT_TMP_PATH}/coredump-timestamp"
```
A single, shared, mutable, global state. All tmt invocations sharing the guest - e.g. with `provision -h local` - will change this location without any kind of synchronization. It's better to use a directory that cannot conflict with anyone else, e.g. AVC check uses `ausearch_timestamp_filepath = invocation.check_files_path / AUSEARCH_TIMESTAMP_FILENAME`.

happz replied:
Plus the `sh -c`, seems to be present in many commands. `ShellScript` will serve us better, `guest.execute(ShellScript(f"mkdir -p {DEFAULT_TMP_PATH} && date '+%Y-%m-%d %H:%M:%S' > {DEFAULT_TMP_PATH}/coredump-timestamp"))`

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,350 @@
+            GuestCapability.CAP_SYS_ADMIN,    # Configure coredump
+            GuestCapability.CAP_SYSLOG,       # Read journal
+            GuestCapability.CAP_DAC_READ_SEARCH,  # Access coredump files
+            GuestCapability.CAP_DAC_OVERRIDE  # Access protected files
```
Why aren't these two capabilities tested in `essential_requires()`?

martinhoyer replied:
Good question - basically this is all a guesswork, I don't actually know if we need these, or any for that matter. Also, I don't think you need `CAP_SYSLOG`, to read journal, at least on my laptop, I have cap_syslog 'not permitted', but can do journalctl, as well as run coredumpctl.

I was experimenting with some ways to figure out if want to skip the coredump check, like
 - doesn't have systemd -> skip
 - doesn't have coredumpctl
   - doesn't have permission to modify coredumpctl config -> skip
   - install coredumpctl
- etc...

But I think I'm overthinking it, imho it can be just has_systemd and can modify the config. Skip everything else, wdyt? (removing essential_requires alltogether)

martinhoyer replied:
simplified in c0c05df176604cf7fa3ec96d92f952e03b886726

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,350 @@
+            self, invocation: "TestInvocation", event: CheckEvent, logger: tmt.log.Logger
+            ) -> ResultOutcome:
+        """Check coredump status and return appropriate result."""
+        # Only configure coredump in before_test
```
Maybe the branch below should be in its own method, e.g. `create_coredump_timestamp()`, which would be then called only before the test. Written like this, the method depends on external state and decides on its own what to do, and if there's a condition like this the outside world knows, why even giving this method a chance to not obey? `before_test` will call it, `after_test` will not.

martinhoyer replied:
Addressed in c0c05df176604cf7fa3ec96d92f952e03b886726, though at some point I became quite confused about what should be where, so reviews needed :)

happz commented:
> @happz continuing the discussion from chat - this is the only solution I found for `GuestCapabilities` and `_query_capabilities`, with which I can sleep at night...I hope it's ok for you too,

As expected, it's a lot of work to get the list of allowed capabilities, but not as much as I was worried. We have our own enum with a lot of entries we don't need, but these new checks start using entries we didn't use before, I guess we may easily find new plugins checking yet another entries in the future. And the list is pretty stable, so LGTM.

> once you check it _after_ your PTO :)

Yeah, yeah, I know, but Geralt needs a break anyway, Ciri is safe, so I have some spare time on my hands.

martinhoyer commented:
> > once you check it _after_ your PTO :)
> 
> Yeah, yeah, I know, but Geralt needs a break anyway, Ciri is safe, so I have some spare time on my hands.

Have you just finished the books or the games? :) 
Thanks for the review! Will try to address most of the comments asap.

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,288 @@
+"""
+
+# Can be set in /proc/sys/kernel/core_pattern
+# CORE_PATTERN = '|/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h'
```
I'm not sure to which this comment belongs.

martinhoyer replied:
Should be gone.

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,288 @@
+                            ignored = True
+                            break
+                    if ignored:
+                        continue
```
If we could live without knowing which pattern matched, this could be a one-liner with `any()`.

martinhoyer replied:
also addressed

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,288 @@
+            return [CheckResult(name="coredump", result=ResultOutcome.SKIP)]
+
+        if not check._create_coredump_timestamp(invocation):
+            outcome = ResultOutcome.ERROR
```
It seems to me this particular change of `outcome` value is ignored, as it can be overwritten by the future call to `check._check_coredump()`.

martinhoyer replied:
hopefully addressed

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,288 @@
+
+    def _check_coredump(
+            self, invocation: "TestInvocation", event: CheckEvent, logger: tmt.log.Logger
+            ) -> ResultOutcome:
```
In the name of connecting dots and letting users in, would it make sense to return not just an outcome, but also the list of logs, i.e. info and coredump files created by `_check_for_crashes()`? I would imagine `_check_for_crashes()` and `_check_coredump()` would basically collect the paths, and add them to `logs` field of the eventual check result, together with the outcome.

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,288 @@
+        # Skip if coredumpctl is not available, as detected in before-test
+        sleep(1)
+        if not check.is_available:
+            return [CheckResult(name="coredump", result=ResultOutcome.SKIP)]
```
I forgot to push it, but, I'm trying to expose as much as possible to end users and CI systems, please, add a note to SKIP outcomes: https://github.com/teemtee/tmt/pull/3498

martinhoyer replied:
Soo, I've rebased it on top of https://github.com/teemtee/tmt/pull/3498 to avoid dealing with it afterwards

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,288 @@
+            outcome = ResultOutcome.ERROR
+
+        check._configure_coredump(invocation.guest, logger)
+        outcome = check._check_coredump(invocation, CheckEvent.BEFORE_TEST, logger)
```
What is expected from checking for coredumps that appear between the timestamp and the beginning of the test?

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -349,34 +394,47 @@ class GuestFacts(SerializableContainer):
     lsb_release_content: dict[str, str] = field(default_factory=dict)
 
-    def has_capability(self, cap: GuestCapability) -> bool:
+    def has_capability(self, capability: GuestCapability) -> bool:
```
Seems like https://github.com/teemtee/tmt/commit/62ad828b667fc6680072f375e6bb30edff9725a9 got lost somewhere during the rebase.

martinhoyer replied:
Indeed. I got tangled in mypy errors. Will get it back.

martinhoyer replied:
ok it's back

happz commented on tmt/steps/provision/podman.py:
```diff
@@ -626,7 +625,7 @@ def go(self, *, logger: Optional[tmt.log.Logger] = None) -> None:
-        self._guest.facts.capabilities[GuestCapability.SYSLOG_ACTION_READ_ALL] = False
-        # ... while this seems to be forbidden completely.
-        self._guest.facts.capabilities[GuestCapability.SYSLOG_ACTION_READ_CLEAR] = False
+    @property
```
This property should not be needed anymore, the one provided by the base class should be doing the very same thing.

martinhoyer replied:
should be gone.

martinhoyer commented:
Added some polish and rebased on main. @happz Can you spot something very bad straight away? fwiw, I'm not super confident about the code, as it's been quite some time and I don't remember everything, on the other hand, there shouldn't be breaking changes, so if @psss, et all wants to have it in the release, maybe as a preview, I can add the docs, etc.

psss commented:
The two red tests seem to be irrelevant failures (failed to boot instance) but the new coredump test seems to be [failing](https://artifacts.osci.redhat.com/testing-farm/811e9932-e9d6-42d5-b295-fa3853c2042d/work-the-restj1xlc9yw/plans/provision/virtual/the-rest/execute/data/guest/default-0/tests/test/check/coredump-15/output.txt) as well:

```
00:00:00     skip coredump (Coredump detection was skipped because coredumpctl is not available or has insufficient privileges.
The coredump check requires systemd-coredump to be installed and the systemd-coredump.socket to be active. Additionally, the user must have sufficient permissions to access coredump data.) (before-test check)
```

martinhoyer commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,412 @@
+                CheckResult(
+                    name="coredump",
+                    result=ResultOutcome.SKIP,
+                    note=[cls._get_not_available_message()],
```
This needs to be improved
```
        note:
          - |-
            Coredump detection was skipped because coredumpctl is not available or has insufficient privileges.

            The coredump check requires systemd-coredump to be installed and the systemd-coredump.socket to be active. Additionally, the user must have sufficient permissions to access coredump data.
            ```

happz replied:
Yes, it does, and maybe it will: https://github.com/teemtee/tmt/pull/3596/files#diff-f9ad5f6605f10698f0208842ba0e42ee7bf1508151827baf78c3660b611e4457R49

martinhoyer replied:
@happz I meant the way I'm adding note here, not the display. Are you saying it's ok here?

happz replied:
I see. I'm not sure what you need to improve then, didn't catch my eye, it will just look ugly, I'm afraid, being longer than just a couple of words.

martinhoyer replied:
Trying this https://github.com/teemtee/tmt/pull/3396/commits/0d506d979a2eb90ef53bfccb3a46719081144134

happz replied:
I wouldn't invest too much work into this particular area; if I get my way with hints and how they are rendered on the console and elsewhere, these messages will get polished anyway to be informative and nice to read. On the other hand, I have no idea when said changes can land in tmt, so maybe it's not a bad idea to have at least something nicer to read.

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,412 @@
+        )
+
+    def _check_for_crashes(
+        self, guest: "Guest", logger: tmt.log.Logger, check_files_path: str, start_time: str
```
A path shall be a `Path`. `invocation.guest, logger, str(invocation.check_files_path), start_time` makes it a string, and than it's used as a string instead of much nicer `pathlib` API: `info_file = check_files_path / f"dump.{exe}_{sig}_{pid}.txt"`, `dump_file = check_files_path / f"dump.{exe}_{sig}_{pid}.core"`

martinhoyer replied:
My bad, thanks! How about https://github.com/teemtee/tmt/pull/3396/commits/f834f2957d640db207ba9a68c5fedfb45093dfc2 ?

happz replied:
WRT

```
                 info_filename = f"dump.{exe}_{sig}_{pid}.txt"
                 info_path = check_files_path / info_filename
```

Would `info_filepath = check_files_path / f"dump.{exe}_{sig}_{pid}.txt"` be too long? I'm not a fan of variables existing just to be used once unless there's some issue with readability or formatting, and `dump_filename` & `info_filename` do not seem to have any other purpose. Also, I'd recommend `*_filepath` naming scheme, I try to use it consistently throughout the tmt codebase, because "file" is quite overloaded, and "filepath" & "dirpath" are cleaner while being just slightly longer.

martinhoyer replied:
Agree, thanks! Will change when near a keyboard

psss replied:
Ok, just for reference: This one has be actually covered: `_check_for_crashes` already uses:

    check_files_path: Path,

And Martin adjusted the `filepath`  naming in 50eb7b616fef03bf1e90c7c6cbfb18b622bb99fe.

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,412 @@
+        # Get list of generated files
+        try:
+            files_output = invocation.guest.execute(
+                Command("find", str(invocation.check_files_path), "-type", "f")
```
I believe `Command` accepts `Path` instances as well as strings. `Command("find", invocation.check_files_path, "-type", "f")` should work like a charm.

martinhoyer replied:
Thought so, but didn't know if it's a good practice, if you say so, I'll do the changes :)

martinhoyer replied:
https://github.com/teemtee/tmt/pull/3396/commits/b03bb6b1725dc88446223529243905558ec1c22a

martinhoyer commented:
> ```
> 00:00:00     skip coredump (Coredump detection was skipped because coredumpctl is not available or has insufficient privileges.
> The coredump check requires systemd-coredump to be installed and the systemd-coredump.socket to be active. Additionally, the user must have sufficient permissions to access coredump data.) (before-test check)
> ```

I presume we don't have permissions to run coredumpctl also on virtual machines - should we?
Adding VMs to "expected skip" and trying to improve the messages: https://github.com/teemtee/tmt/pull/3396/commits/c69e2657b474a787f56355f97da862fb2c7a05cc https://github.com/teemtee/tmt/pull/3396/commits/0d506d979a2eb90ef53bfccb3a46719081144134

martinhoyer commented:
> I presume we don't have permissions to run coredumpctl also on virtual machines - should we?  

hmm, maybe that not a good presumption. Will investigate...

happz commented:
> > I presume we don't have permissions to run coredumpctl also on virtual machines - should we?
> 
> hmm, maybe that not a good presumption. Will investigate...

Yeah, running `coredumpctl` in a full-fledged VM, as `root`, that does sound like something we should be able to do. If I'm not mistaken, there is no issue with AVC check in the same environment.

martinhoyer commented:
```
[root@default-0 tree]# systemctl status systemd-coredump.socket
● systemd-coredump.socket - Process Core Dump Socket
     Loaded: loaded (/usr/lib/systemd/system/systemd-coredump.socket; static)
     Active: active (listening) since Wed 2025-03-26 15:34:30 UTC; 1min 38s ago
 Invocation: 7dbf694668af40efab524431d094d022
       Docs: man:systemd-coredump(8)
     Listen: /run/systemd/coredump (SequentialPacket)
   Accepted: 0; Connected: 0;
     CGroup: /system.slice/systemd-coredump.socket
 ```
 coredump seems to be up on a vanilla `Fedora-Cloud-Base-Generic` VM for me. Maybe it was caused by the unrelated failures you mentioned @psss ?

martinhoyer commented:
oh lord:
```
# coredumpctl list --no-pager
No coredumps found.
[root@vm-10-0-184-99 ~]# echo $?
1
```
```
def _has_required_permissions(self, guest: "Guest", logger: tmt.log.Logger) -> bool:
        """Check if we have sufficient permissions to access coredump data."""
        try:
            guest.execute(ShellScript("coredumpctl list --no-pager")
```

martinhoyer commented:
F40 run does not seem to be related:
```
[   FAIL   ] :: File '/var/tmp/rlRun_LOG.bTPO2w8p' should contain 'PLAY \[Enable EPEL repositories\]' 
[   FAIL   ] :: File '/var/tmp/rlRun_LOG.bTPO2w8p' should contain 'Run .* successfully finished. Bye for now!' 
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::   RESULT: FAIL (Epel)
```
`tests/try/basic/Epel`

psss commented:
Seems the problem is still [there](https://artifacts.osci.redhat.com/testing-farm/02723c7e-1839-4520-93c2-d19e68a68196/work-the-restl4nwus93/plans/provision/virtual/the-rest/execute/data/guest/default-0/tests/test/check/coredump-15/output.txt):

```
pass /coredump/harmless
    output.txt: /tmp/tmp.LysbJDVUpn/plan/execute/data/guest/default-0/coredump/harmless-1/output.txt
    skip coredump (Coredump detection was skipped: insufficient permissions to access coredump data.

    The coredump check requires systemd-coredump to be installed and the systemd-coredump.socket to be active. Additionally, the user must have sufficient permissions to access coredump data.) (before-test check)
    skip coredump (Coredump detection was skipped: insufficient permissions to access coredump data.

    The coredump check requires systemd-coredump to be installed and the systemd-coredump.socket to be active. Additionally, the user must have sufficient permissions to access coredump data.) (after-test check)
```

martinhoyer commented:
Hopefully complete now.  

I've added `match-patterns` feature to a different branch - https://github.com/martinhoyer/tmt/compare/coredump...teemtee:tmt:coredumo-match-patterns 
To be implemented once we have a good way to run multiple checks with the same name.

martinhoyer commented on tests/test/check/test-coredump.sh:
```diff
@@ -74,7 +74,7 @@ rlJournalStart
 
         # This test should pass regardless of systemd availability
-        rlRun "tmt run --id $run --scratch -a -vv provision -h $PROVISION_HOW test -n /coredump/ignore-pattern"
+        rlRun "tmt run --id $run --scratch -a -vv provision -h $PROVISION_HOW test -n /coredump/ignore-pattern -ddddvvvv"
```
argh, a leftover from the endless pain and suffering.

martinhoyer commented on tmt/checks/coredump.py:
```diff
@@ -147,32 +147,28 @@ def _has_active_socket(self, guest: "Guest", logger: tmt.log.Logger) -> bool:
         # Try without sudo first
         try:
-            guest.execute(ShellScript("coredumpctl list --no-pager"), silent=True)
+            guest.execute(ShellScript("coredumpctl info &>/dev/null || [ $? -eq 1 ]"), silent=True)
```
I probably should have added extra comment. It is my understanding that returncode 1 is safe to use as any permission or other wierd problems would return a different code.  

That said, I gave up trying to decipher https://github.com/systemd/systemd/blob/main/src/basic/errno-util.h after 5 seconds.

martinhoyer commented on tmt/checks/coredump.py:
```diff
@@ -218,6 +214,49 @@ def _check_coredump_available(
 
         return True, ""
 
+    def _wait_for_coredump_processes(self, guest: "Guest", logger: tmt.log.Logger) -> None:
```
Would be cool to have a test for this, but I'm definitely gonna be able to find better things to do in near future.

martinhoyer commented on spec/tests/check.fmf:
```diff
@@ -10,9 +10,9 @@ description: |
-    the user can run into. Another useful checks would be kernel
-    panic detection, core dump collection or collection of system
-    logs.
+    the user can run into. Other useful checks include kernel panic
```
Well, this might not fly, but it technically detects kernel panic :lab_coat:

happz commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,601 @@
+    """
+
+    @staticmethod
+    def _get_not_available_message(reason: str = "") -> str:
```
I'd recommend using `None` instead of `""` - it is much easier for type checkers to follow `None` around, while `""` is a valid string. For signaling "unset" or "unknown", `None` is much better choice. Applies to `is_availability_reason` too, it should be `Optional[str]`.

martinhoyer replied:
Addressed

psss replied:
This one seems to be addressed.

psss commented on spec/tests/check.fmf:
```diff
@@ -74,9 +74,18 @@ example:
+    # Configure the coredump check with ignore patterns
+    check:
+      - how: coredump
+        ignore-patterns:
```
```suggestion
        ignore-pattern:
```

Let's use singular to be consistent with other keys.

psss commented on tests/test/check/data/main.fmf:
```diff
@@ -61,6 +61,32 @@
+    tag+:
+      - provision-only
+      - provision-container
+      - provision-virtual
```
This tags are only used for the top-level test selection, not used under the `data` directories.

psss replied:
Thanks!

psss commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,601 @@
+                if self.ignore_patterns and any(
+                    pattern.search(crash_info) for pattern in self.ignore_patterns
+                ):
+                    logger.debug(f"Ignoring crash due to pattern match in: {crash_info}")
```
Would be nice to include a `note` that crash happened but was ignored based on the `ignore-pattern` key.

happz replied:
Please, add a subissue to https://github.com/teemtee/tmt/issues/3553, that's right in the area.

psss replied:
Is this supposed to be a hint? This is just a regular check result note, something similar to `test raised avc as expected` here we get `crash detected but ignored as instructed`.

happz replied:
Right, I wouldn't make it a hint, it's just the area I attempt to poke, i.e. make very trivial messages like "timeout" more verbose and informative. This covers not just hints, but also result notes and exceptions.

Feel free to file a standalone issue, if you do so, please refer to #3553 so I can keep track of what's left.

psss replied:
Ok, filed as #3653.

martinhoyer replied:
As discussed, adding a note wouldn't be trivial here. Can be done in other PR, alongside the other TODOs waiting for #3498.
I've added there at least: `logger.info( f"Ignoring crash in PID {pid} ({exe}) due to pattern match: '{matching_pattern.pattern}'"`

martinhoyer commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,601 @@
+                if self.ignore_patterns and any(
+                    pattern.search(crash_info) for pattern in self.ignore_patterns
+                ):
+                    logger.debug(f"Ignoring crash due to pattern match in: {crash_info}")
```
Add a note to CheckResult saying it was matched and switch from .debug to .info

therazix commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,614 @@
+                guest.execute(
+                    ShellScript("sudo mkdir -p /etc/systemd/coredump.conf.d")
+                    & ShellScript(
+                        f"echo '{COREDUMP_CONFIG}' | sudo tee"
```
Missing space. Without it, the script would look like this: `...sudo tee/etc/...`
```suggestion
                        f"echo '{COREDUMP_CONFIG}' | sudo tee "
```

martinhoyer replied:
fixed, thanks!

psss commented:
> Addressed feedback, rebased on main.

Thanks for the update. Next time, could you please add new commits instead of amending and force-pushing? It helps with the review to see what has changed, especially when combined with rebase which brings other irrelevant changes so even github's `martinhoyer force-pushed` links do not help.

psss commented on tests/test/check/data/main.fmf:
```diff
@@ -61,6 +61,28 @@
+    /ignore-pattern:
+        check:
+          - how: coredump
+            ignore-patterns:
```
Here the plural form is still used and the test has passed? Seems the implementation has not been adjusted accordingly.

martinhoyer replied:
updated

psss replied:
Seems to be fixed now. Thanks.

therazix commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,614 @@
+            return False
+
+    def _check_coredump(
+        self, invocation: "TestInvocation", event: CheckEvent, logger: tmt.log.Logger
```
The `event` parameter is not used anywhere. Is there a reason to keep it?

therazix replied:
Do we need it? The function is called only from `before_test` and `after_test`. And I don't see anything that says that the `_check_coredump` function must have the `event` parameter, so you can remove it from the call too. Or am I missing something?

therazix commented on tmt/checks/coredump.py:
```diff
@@ -0,0 +1,614 @@
+        # from tmt.utils.hints import get_hints
+
+        # Skip if coredumpctl is not available, as detected in before-test
+        sleep(1)
```
Forgotten `sleep`?

martinhoyer replied:
Ha! Correct :)

martinhoyer commented:
> Next time, could you please add new commits instead of amending and force-pushing? It helps with the review to see what has changed, especially when combined with rebase which brings other irrelevant changes so even github's martinhoyer force-pushed links do not help.

@psss 
Yep, also not happy how it turned out, apologies. I thought that if I first add separate commits, push them and do the rebase+push it will look decent, but alas it's broken. I learned my lesson :saluting_face:
